[superb-sushi] iP

superb-sushi opened the pull request:
### Blob (your next task chatbot)

> Don't be a snob with your todo lists! Use Blob!

- text-based
- fast to use
- intuitive!

**All you need to do is...**
1. Download the .jar file from [here](https://github.com/superb-sushi/ip/releases)
2. Save it in a folder
3. Navigate to that exact directory using your command prompt
4. Enter `java -jar ip.java`
5. Add your tasks
6. Get carried away!

And it is ~~$10.00~~ **FREE**! 😮 
___
Features:
- [x] Managing Tasks
- [ ] Managing Deadlines (coming soon)
- [ ] Reminders (coming soon)
___
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method!
``` Java
public class Blob {
  public static void main(String[] args) {
          new Blob("./database.csv").run();
      }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

superb-sushi committed:
Level-0

superb-sushi committed:
Level-1

superb-sushi committed:
Level-2

superb-sushi committed:
Level-3

superb-sushi committed:
Level-4

superb-sushi committed:
A-TextUiTesting

superb-sushi committed:
Level-5

superb-sushi committed:
Level-6

superb-sushi committed:
Add 'Save' features to existing project

Project supports simple CRUD functionalities, but the changes made to the listed tasks are not permanently saved onto the hard drive of the user's device.

This prevents previous changes made to the task list from being 'remembered' by the program, rendering the program useless should the user wish to save past tasks.

This commit introduces two new methods:
* getFileContents
* updateFileContents
getFileContents retrieves the information of all previously made tasks (todos, deadlines, and events) from a CSV file (which serves as the database) and returns an ArrayList<Task> containing all these tasks, allowing for the program to interact with past tasks.
updateFileContents iterates through the ArrayList<Task> and writes each and every task onto the CSV file upon being called. It is placed at the end of every CRUD method found in the program.

Now the program is able to output previously created tasks through reading the contents of the 'database.csv' file!

superb-sushi committed:
Merge branch 'master' into branch-Level-7
Making 'branch-Level-7' the same with 'master'

superb-sushi committed:
Add DateTime Optimisation to Tasks

Current tasks utilise Strings to represent date and time of 'event' and 'deadline' tasks, which are not as ideal as using LocalDateTime data types.

The relevant fields have been changed to store LocalDateTime types, allowing for clearer and standardised storage of timings and printing of deadlines.

superb-sushi committed:
Add more OOP features to source code

All of the code is currently stored in a single class 'Blob' with the use of nested classes (e.g. Task, Deadline, Todo...).

Separation of Nested Classes into separate files. Creation of more classes such as Ui, Storage, Parser and TaskList.

Use of separate files for the different varying classes makes it much easier for the developer to isolate each portion of the code. Addition of more classes (such as Ui, Storage, Parser, and TaskList) help to abstract away more code for easier understanding and debugging.

superb-sushi committed:
Organise all classes in package blob

Program cannot be imported out as it is not packaged

All classes are not organised under a singular package, 'blob'.

Now program can be imported.

superb-sushi committed:
Merge remote-tracking branch 'origin/add-gradle-support'
Merging origin/add-gradle-support into master

superb-sushi committed:
Add JUnit tests TaskListTest & StorageTest

No automated tests initially to test certain classes and their methods.

This makes it harder to isolate bugs should the program run into an error.

Introducing StorageTest and TaskListTest helps to isolate bugs that may arise due to issues with the code in TaskList and Storage.

superb-sushi committed:
Minor Changes to 'build.gradle' file

superb-sushi committed:
Add JavaDocs to all classes and methods

Existing Classes and Methods do not have headers explaining what they do and their role.

JavaDocs was added to all classes and methods.

This allows the developer to have greater clarity when it comes to remembering what each and every class / method does, expediting the programming process.

superb-sushi committed:
Merge branch 'branch-Level-7'
For submission tracking purposes...

superb-sushi committed:
Refactor Source Code to meet Coding Standards

Current code doesn't follow coding standards, thus impeding clarity when scanning through the multiple files, potentially leading to slower bug finding efforts.

Changing of code to abide by the coding standards allows for consistent style throughout, allowing for smooth debugging / project building progress.

superb-sushi committed:
Add 'find' command word

User can only 'list' out all existing tasks to see more than one task that is currently recorded in the database.

Lack of a search functionality renders the user unable to sieve out specific tasks, especially once more and more tasks get added to the database.

The adding of the 'find' command allows the user to sift out specific tasks that contain that keyword, allowing for improved user experience.

superb-sushi committed:
Integrate JavaFX GUI to Blob

Blob is currently using a text-based UI.

Integrating a GUI makes Blob easier to use and more engaging.

tckeong commented on src/main/java/blob/DialogBox.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public static DialogBox getBlobDialog(String text, Image img) {
+        var db = new DialogBox(text, img);
```
I think here can change the variable name db to something is more relevant, like dialogBox. Since db maybe confusing as database.

tckeong commented on src/main/java/blob/Event.java:
```diff
@@ -0,0 +1,43 @@
+     */
+    @Override
+    public String toString() {
+        return "[E]" + "[" + this.check() + "] " + this.name + " (from: " + start.format(DateTimeFormatter.ofPattern("MMM d HH:mm")) + " to: " + end.format(DateTimeFormatter.ofPattern("MMM d HH:mm")) + ")";
```
According to the [JavaDocs](https://se-education.org/guides/conventions/java/intermediate.html), the indentation for wrapped lines should be 8 spaces .

SKarthikeyan28 replied:
Could add break after "(from:" to improve readability?

tckeong commented on src/main/java/blob/Parser.java:
```diff
@@ -0,0 +1,258 @@
+ */
+public class Parser {
+
+    /**
+     * @param ui Ui used by the chatbot.
+     * @param tasklist Tasklist containing all exisitng tasks.
+     * @param action String representation of the input from the user.
+     * Existing accepted commands now include:
+     *               'bye' - exits the conversation with the bot
+     *               'list' - lists all existing tasks
+     *               'mark X' - marks X'th task as done
+     *               'unmark X' - marks X'th task as not done
+     *               'delete X' - removes X'th task from the database
+     *               'todo task' - adds a to-do task to database
+     *               'deadline task /by YYYY-MM-DD HH:mm' - adds a deadline task to the database
+     *               'event task /from YYYY-MM-DD HH:mm /to YYYY-MM-DD HH:mm' - adds a event task to the database
+     */
```
Maybe can add the description for this function.

tckeong commented on src/main/java/blob/Parser.java:
```diff
@@ -0,0 +1,258 @@
+                }
+
+                // for task deadline string
+                StringBuilder sDate = new StringBuilder(arr[by + 1]);
+                StringBuilder sTime = new StringBuilder(arr[by + 2]);
```
Maybe can make the name become more explainable to what it do, like sDate can become strDate or stringDate

tckeong commented on src/main/java/blob/Parser.java:
```diff
@@ -0,0 +1,258 @@
+                }
+
+                //for start string
+                StringBuilder stDate = new StringBuilder(arr[start + 1]);
+                StringBuilder stTime = new StringBuilder(arr[start + 2]);
```
Maybe can make the name become more explainable to what it do, like stDate can become startDate, stTime can become startTime

tckeong commented on src/main/java/blob/Parser.java:
```diff
@@ -0,0 +1,258 @@
+                String startISO8601Format = stDate.toString() + "T" + stTime.toString() + ":00";
+
+                //for end string
+                StringBuilder enDate = new StringBuilder(arr[end + 1]);
+                StringBuilder enTime = new StringBuilder(arr[end + 2]);
```
Maybe can make the name become more explainable to what it do, like enDate can become endDate, and enTime can become endDate

tckeong commented on src/main/java/blob/Storage.java:
```diff
@@ -0,0 +1,95 @@
+ * Constructor is to be called with a string representation of the desired file path to store the database as its argument.
+ */
+public class Storage {
+    private File f;
```
Maybe can make the name become more explainable to what it do, like f become file

tckeong commented on src/main/java/blob/Task.java:
```diff
@@ -0,0 +1,46 @@
+package blob;
```
Can organize all class relevant to tasks into one package.

tckeong commented on src/main/java/blob/TaskList.java:
```diff
@@ -0,0 +1,111 @@
+     */
+    public Task getTask(int index) {
+        try {
+            ArrayList<Task> db = this.storage.getFileContents();
```
Maybe can change the db to a name more explainable.

tckeong commented on src/main/java/blob/Ui.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class Ui {
+
+    public final String unknownCommandMsg = "ERROR! Unknown command!";
```
The constant should be in all uppercase using underscore to separate words (aka SCREAMING_SNAKE_CASE).

tckeong reviewed (commented):
Overall are good, only some minor code styles stuffs and the naming stuffs.

SKarthikeyan28 commented on src/main/java/blob/Blob.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    /**
+     * Kicks start the chatbot's chatting functionality
```
Could add param to javadoc

SKarthikeyan28 commented on src/main/java/blob/Deadline.java:
```diff
@@ -0,0 +1,31 @@
+     */
+    @Override
+    public String toString() {
+        return "[D]" + "[" + this.check() + "] " + this.name + " (by: " + deadline.format(DateTimeFormatter.ofPattern("MMM d HH:mm")) + ")";
```
Could add break after "by:" to improve readability?

SKarthikeyan28 commented on src/main/java/blob/Storage.java:
```diff
@@ -0,0 +1,95 @@
+                
+            } else if (t.type.equals("E")) {
+                Event e = (Event) t;
+                StringBuilder data = new StringBuilder(String.format("%s,%s,%s,%s,%s\n", t.type, isChecked, t.name, e.getStart(), e.getEnd()));
```
Could add break after comma to improve readability

SKarthikeyan28 commented on src/main/java/blob/TaskList.java:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * @return integer number of existing tasks.
+     */
+    public int size() {
```
Maybe can change method name to be a verb like "getSize"

SKarthikeyan28 reviewed (commented):
Overall code looks good, adheres to java coding style. **LGTM**!!

superb-sushi committed:
Add assertions to Storage, Deadline and Event files

CSV file for data storage is assumed to already have the proper headers, while the deadlines and relevant time periods entered by the user are assumed to be logical timings for both the deadline and event tasks created.

Inappropriate dates entered may render the tasks created useless and of little utility to the user, and may perhaps interfere with future iterations to the project which leverage on the created tasks having proper and logical dates.

Assertions have been added to ensure only deadlines past the time the user creates the task can be used. Same goes with the duration of event tasks as indicated by the user.

superb-sushi committed:
Merge pull request #1 from superb-sushi/branch-A-Assertions

Add assertions to Storage, Deadline and Event files

superb-sushi committed:
Adherence to Coding Quality

Non-standardised variable namings present throughout codespace.

Changing of variable names to adhere to coding quality, such as giving names that more intuitively describe what the variable holds, making method names verbs, prevention of misleading variable names.

Doing as such allows for greater readability and clarity when analysing the codebase for future bugs. Also allows for more expedient understanding when code is read by other users.

superb-sushi committed:
Merge pull request #2 from superb-sushi/branch-A-CodeQuality

Adherence to Coding Quality

superb-sushi committed:
Add Tagging Feature to all tasks

Current tasks are differentiated by types (Todo, Deadline, Event), but within each subtype, there isn't any other way to differentiate the tasks into smaller subgroups.

With the addition of the tagging feature, tasks now can be classified under different categories as determined by the user (e.g. #academics), with each task being able to have an unlimited number of tags.

Now the user would be able to visually group related tasks (which are associated with the same tags) together, adding greater utility to Blob as a task tracking app.

superb-sushi committed:
Merge pull request #3 from superb-sushi/branch-C-Tagging

Add Tagging Feature to all tasks

superb-sushi committed:
Refactor of Code to fit Coding Standards

The evaluateAction() method in Parser is extremely long.

The length of the method obscures the clarity of what it actually does
& makes the rectifying of bugs less expedient.

Let's separate all the various commands into separate methods to leverage
on the SLAP method, while keeping the method short and concise.

superb-sushi committed:
Add Ui.png and User Guide

Lack of relevant User guide and picture of messenger interface.

The product is less intriging to potential users due to the lack
of visual aids and a consumer-friendly explanation of what the
product does.

Let's add a .png file showcasing the first thing the user sees
when they run the product. A proper User Guide also helps users
to better understand how to use / navigate the product!

superb-sushi committed:
Update Ui.png

Ui.png had a lot of blank areas, making the product appear
less-populated and thus, less interesting.

Let's populate the interface with actual uses of the
built-in commands to make it more engaging, as well as
to highlight it's efficacy!

superb-sushi committed:
Alter database path for JAR

Database creation / access path is based off the layout in IntelliJ

The many different folders and directory's in the layout of Intelllij
mis different from that when the user downloads the JAR file into their
own folder, making it vulnerable to errors when run as a JAR file in a
folder that does not have the same layout.

Let's change the path to make it not dependent on other folders, and
just have it created in the same directory as the JAR file.

superb-sushi committed:
Implement 'switch-case' statements

The updateFileContents() method as well as the getFileContents()
method utilised if-else statements

This  makes the code less readable given how it makes the code
more nested.

Let's change these if-else statements to 'switch' statements
instead, making the code more palatable to the eye.

superb-sushi committed:
Update User Guide

superb-sushi committed:
Merge pull request #5 from superb-sushi/branch-UG

Update User Guide

superb-sushi committed:
Update User Guide Again

superb-sushi committed:
Merge pull request #6 from superb-sushi/branch-UG

Update User Guide Again

superb-sushi committed:
Update User Guide

superb-sushi committed:
Merge pull request #7 from superb-sushi/branch-UG

Update User Guide

superb-sushi committed:
Update User Guide Again

superb-sushi committed:
Merge pull request #8 from superb-sushi/branch-UG

Update User Guide Again

superb-sushi committed:
Debug Parser.java

superb-sushi committed:
Merge pull request #9 from superb-sushi/branch-W8-updates

Debug Parser.java
